### ISSUE TITLE
Add partitions after stream is created

### ISSUE THREAD

**Issue by hyprnick** [liftbridge-io/liftbridge#279]
Is it possible to add more partitions after a stream is created?

**Comment by tylertreat** [comment]
Not currently, no. The challenge with adding partitions or, more generally, re-partitioning (adding or removing partitions) a stream is that it requires re-partitioning all of the messages in the stream since they may now map to different partitions. This basically requires rewriting all of the messages to the re-partitioned stream using the appropriate client-side partitioning logic, so it's not a trivial thing.

**Comment by hyprnick** [comment]
Thanks for the help! 

What about only allowing someone to add partitions and not remove any?  This is how Kafka works.  New messages don't immediately go to the new partitions unless you update your hashing logic to use the new partitions.  

If this isn't possible, would a potential workaround be to create a new stream or would this cause too much overhead on the cluster?  

This is a great project, thank you for working on this.  Hope to make contributions in the future.

**Comment by tylertreat** [comment]
> What about only allowing someone to add partitions and not remove any? This is how Kafka works. New messages don't immediately go to the new partitions unless you update your hashing logic to use the new partitions.

This would definitely be feasible though not currently possible. I'd need to think about how this would work on the client side as far as updating hashing logic (curious how Kafka handles this).

> If this isn't possible, would a potential workaround be to create a new stream or would this cause too much overhead on the cluster?

This would be my recommendation. You could always create streams with a single partition as needed and handle the partitioning logic yourself.

I would definitely like to explore options for providing a nicer re-partitioning story and very open to suggestions.

**Comment by hyprnick** [comment]
I wouldn't know exactly how Kafka handles it but based on observation, the round robin partitioner starts to use the new partitions once the metadata is updated.  For a custom partitioner, that's all created client side so if the client is updated with the new set of partitions, (say a mod total), then it would use them.  

That was my next question - how much overhead is a stream with a single partition vs using multiple partitions?  

At a previous engagement, rebalancing took days so we tried to avoid it at all costs.  Constantly trying to fix hot partitions was a lot of hand-holding.  That would be an interesting problem to work on.

**Comment by tylertreat** [comment]
> the round robin partitioner starts to use the new partitions once the metadata is updated.

It seems like this could be problematic depending on how data is partitioned. E.g. if you're partitioning on, say, a user id such that all events for the same user go to the same partition and you add partitions, eventually you'll end up with two or more partitions containing a given user's events.

> That was my next question - how much overhead is a stream with a single partition vs using multiple partitions?

There is no overhead. In fact, partitions didn't originally exist in Liftbridge, just streams. At that time, it was the client's responsibility to handle partitioning over streams. Partitions were introduced to provide a better user experience. Basically what that entailed was changing "stream" to be "partition."  So a stream is effectively just a logical wrapper around a collection of partitions. All of the overhead is at the partition level, so a stream with 30 partitions is roughly equivalent to 30 streams with 1 partition in terms of resources.

> At a previous engagement, rebalancing took days so we tried to avoid it at all costs. Constantly trying to fix hot partitions was a lot of hand-holding. That would be an interesting problem to work on.

Yep, I would really like to make this a nicer experience than what Kafka offers (since that is generally the goal of Liftbridge), but it's a challenging problem.

**Comment by strahe** [comment]
Just want to know if this feature is being considered?
I think this is a very useful feature.

**Comment by tylertreat** [comment]
@strahe I agree reconfiguring partitions is useful and is absolutely worth considering. However, it's not yet clear what the path forward should be. I believe allowing partitions to be added without re-mapping messages in existing partitions is problematic as it potentially breaks partitioning semantics (as mentioned in the example above).

The "correct" solution seems to be replaying the existing messages and reapplying the partitioning logic to map them to the correct partition. There are a couple challenges with this though. First, unless it is using built-in partitioners (partition by round robin or by message key), the partitioning logic is custom to the client, so that would somehow need to be supplied to the server (if re-partitioning is to be handled by the server). Second, this re-partitioning would be an expensive operation to perform on the server and would likely require the stream to not accept new messages until the re-partitioning is completed, which causes availability concerns.

Suffice to say, there needs to be more thought to how re-partitioning a stream should work, what the acceptable trade-offs are, etc.
